### PR TITLE
Fixing the race condition when caching GSS contexts

### PR DESCRIPTION
**Pull request by srirampatil**
GSS_DATA and GSS_DESTROY requests expect the context is prensent in the
cache, otherwise it returns RPCSEC_GSS_CREDPROBLEM. The GSS context is
established and cached in GSS_INIT/GSS_CONTINUE_INIT requests. NTIRPC
sends a reply back to the client before inserting the context in cache.
This fix addresses the race and makes sure that the GSS context is
inserted before sending the response to client.

### CODE REVIEW

File: src/svc_auth_gss.c [hunk 1]
```diff
@@ -544,6 +548,7 @@ _svcauth_gss(struct svc_req *req, bool *no_dispatch)
 
 		if (call_stat >= XPRT_DIED) {
 			rc = AUTH_FAILED;
+			(void)authgss_ctx_hash_del(gd);
 			goto gd_free;
 		}
```
Comment (dang) [inline src/svc_auth_gss.c:555]: I wonder if this whole block is necessary.  At least established, I think.  Maybe just move the whole block above the reply?

Comment (srirampatil) [reply src/svc_auth_gss.c:555]: The block actually modifies gr.gr_major when it calls gss_get_name_attribute. And also sets the MSPAC flags in gd. So, I am not sure what effect is it going to have on the returned value. Maybe we can set established before sending the reply.

**Comment by srirampatil** [comment]
@dang Can you please take a look at these changes? I have responded to your previous comment.

**Review by dang — APPROVED** [review]

**Comment by dang** [comment]
Can you please sign your commit?  It can be done with:
git commit --amend -s

**Comment by srirampatil** [comment]
> Can you please sign your commit? It can be done with:
> git commit --amend -s

Done.

**Pull request merged**